dll/win32/browseui/shellbrowser.cpp

```diff
@@ -4086,6 +4165,51 @@ LRESULT CShellBrowser::OnAppCommand(UINT uMsg, WPARAM wParam, LPARAM lParam, BOO
     return 0;
 }
 
+LRESULT CShellBrowser::OnToggleFullscreen(WORD wNotifyCode, WORD wID, HWND hWndCtl, BOOL &bHandled)
+{
+    enum { SF_MBAR = 0x1, SF_SBAR = 0x2, SF_RESIZE = 0x4, SF_MAXIMIZED = 0x8 };
+    const UINT OrgUiSetAuto = m_BrowserSvcFlags & BSF_UISETBYAUTOMATION;
+    const BOOL fCurrentlyFullscreen = (m_BrowserSvcFlags & BSF_THEATERMODE);
+    const BOOL fEnter = wID ? !fCurrentlyFullscreen : wNotifyCode;
+    const HWND hWnd = GetTopLevelBrowserWindow();
+    if (fEnter)
+    {
+        VARIANT_BOOL varb;
+        m_NonFullscreenState = (m_BrowserSvcFlags & BSF_RESIZABLE) ? SF_RESIZE : 0;
+        m_NonFullscreenState |= (FAILED(get_MenuBar(&varb)) || varb) ? SF_MBAR : 0;
+        m_NonFullscreenState |= (FAILED(get_StatusBar(&varb)) || varb) ? SF_SBAR : 0;
+        m_NonFullscreenState |= (SHSetWindowBits(hWnd, GWL_STYLE, 0, 0) & WS_MAXIMIZE) ? SF_MAXIMIZED : 0;
+        SetFlags(BSF_THEATERMODE, BSF_THEATERMODE);
+        put_MenuBar(VARIANT_FALSE);
+        put_StatusBar(VARIANT_FALSE);
+        SHSetWindowBits(hWnd, GWL_STYLE, WS_CAPTION | WS_BORDER | WS_DLGFRAME | WS_THICKFRAME, 0);
+        SHSetWindowBits(hWnd, GWL_EXSTYLE, WS_EX_WINDOWEDGE, 0);
+        ::ShowWindow(hWnd, SW_SHOWMAXIMIZED);
+        MONITORINFO mi;
+        GetWindowMonitorInfo(hWnd, mi);
+        int x = mi.rcMonitor.left, w = mi.rcMonitor.right - x;
+        int y = mi.rcMonitor.top, h = mi.rcMonitor.bottom - y;
+        ::SetWindowPos(hWnd, HWND_TOPMOST, x, y, w, h, SWP_FRAMECHANGED);
+        ::RedrawWindow(hWnd, NULL, NULL, RDW_FRAME | RDW_INVALIDATE | RDW_UPDATENOW | RDW_ALLCHILDREN);
+    }
+    else
+    {
+        SetFlags(0, BSF_THEATERMODE);
+        put_MenuBar((m_NonFullscreenState & SF_MBAR) ? VARIANT_TRUE : VARIANT_FALSE);
+        put_StatusBar((m_NonFullscreenState & SF_SBAR) ? VARIANT_TRUE : VARIANT_FALSE);
+        SHSetWindowBits(hWnd, GWL_EXSTYLE, WS_EX_WINDOWEDGE, WS_EX_WINDOWEDGE);
+        UINT styles = WS_CAPTION | WS_BORDER | WS_DLGFRAME | ((m_NonFullscreenState & SF_RESIZE) ? WS_THICKFRAME : 0);
+        SHSetWindowBits(hWnd, GWL_STYLE, styles | WS_THICKFRAME, styles);
+        ::SetWindowPos(hWnd, HWND_TOP, 0, 0, 0, 0, SWP_FRAMECHANGED | SWP_NOMOVE | SWP_NOSIZE);
+        ::ShowWindow(hWnd, SW_SHOWNOACTIVATE);
+        if (m_NonFullscreenState & SF_MAXIMIZED)
+            ::ShowWindow(hWnd, SW_SHOWMAXIMIZED);
+        ::RedrawWindow(hWnd, NULL, NULL, RDW_FRAME | RDW_INVALIDATE | RDW_UPDATENOW | RDW_ALLCHILDREN);
```

Comment:
Could be moved down outside the if-else blocks, since it's common between the two.